silkworm/db/snapshots/index/btree_index.cpp

```diff
@@ -0,0 +1,151 @@
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "btree_index.hpp"
+
+#include <utility>
+
+#include <gsl/util>
+
+#include <silkworm/core/common/endian.hpp>
+#include <silkworm/core/common/util.hpp>
+#include <silkworm/infra/common/ensure.hpp>
+
+namespace silkworm::snapshots::index {
+
+BTreeIndex::BTreeIndex(seg::Decompressor& kv_decompressor,
+                       std::filesystem::path index_file_path,
+                       std::optional<MemoryMappedRegion> index_region,
+                       uint64_t btree_fanout)
+    : file_path_(std::move(index_file_path)) {
+    ensure(kv_decompressor.is_open(), "BTreeIndex: KV file decompressor must be opened");
+
+    // Gracefully handle the case of empty index file before memory mapping to avoid error
+    if (std::filesystem::file_size(file_path_) == 0) {
+        return;
+    }
+
+    // Either use given memory-mapped region or create a new one
+    memory_file_ = std::make_unique<MemoryMappedFile>(file_path_, index_region);
+    SILKWORM_ASSERT(memory_file_->size() > 0);
+    const auto memory_mapped_range = memory_file_->region();
+
+    // Read encoded Elias-Fano 32-bit list of integers representing data offsets
+    data_offsets_ = EliasFanoList32::from_encoded_data(memory_mapped_range);
+    ensure(data_offsets_->sequence_length() > 0, "BTreeIndex: invalid zero-length data offsets");
+
+    const auto encoded_nodes = memory_mapped_range.subspan(data_offsets_->encoded_data_size());
+
+    // Let the OS know we're going to read data sequentially now, then restore normal (i.e. unknown) reading behavior
+    kv_decompressor.advise_sequential();
+    [[maybe_unused]] auto _ = gsl::finally([&]() { kv_decompressor.advise_normal(); });
+    auto kv_it = kv_decompressor.begin();
+
+    btree_ = std::make_unique<BTree>(
+        data_offsets_->sequence_length(),
+        btree_fanout,
+        [this](auto data_index, auto& data_it) { return lookup_data(data_index, data_it); },
+        [this](auto key, auto data_index, auto& data_it) { return compare_key(key, data_index, data_it); },
+        kv_it,
+        encoded_nodes);
+}
+
+BTreeIndex::Cursor::Cursor(BTreeIndex* index, ByteView key, ByteView value, DataIndex data_index, DataIterator data_it)
+    : index_(index), key_(key), value_(value), data_index_(data_index), data_it_(std::move(data_it)) {}
+
+std::optional<BTreeIndex::Cursor> BTreeIndex::seek(ByteView seek_key, DataIterator data_it) {
+    const auto [found, key, value, data_index] = btree_->seek(seek_key, data_it);
+    if (key.compare(seek_key) >= 0) {
+        return new_cursor(key, value, data_index, data_it);
+    }
+    return std::nullopt;
+}
+
+std::optional<Bytes> BTreeIndex::get(ByteView key, DataIterator data_it) {
+    if (empty()) {
+        return {};
```

Comment:
I prefer `std::nullopt` to be more explicit (you also use it above)

Reply:
We don't have a common convention for this (e.g. in `core` package we use `{}` for such cases), but I agree 100% with you to change here because we should at least be consistent in the same file.